### PR TITLE
JFrog Xray parser

### PR DESCRIPTION
**Pull request by madchap**
**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Your code is flake8 compliant 
- [x] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [ ] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [x] Add applicable tests to the unit tests.

### CODE REVIEW

File: dojo/tools/jfrogxray/parser.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+                unique_key = node.get('id') + node.get('summary') + node.get('provider') + node.get('source_comp_id') + \
+                    title_cve
+                items[unique_key] = item
+
+        return list(items.values())
+
+
+def get_item(vulnerability, test):
+    print(vulnerability)
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]: Maybe use a logger here?

File: dojo/tools/jfrogxray/parser.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    finding = Finding(
+        title=title,
+        cve=cve,
+        cwe=cwe,
+        test=test,
+        severity=severity,
+        description=vulnerability['summary'] + "\n\n" + extra_desc,
+        mitigation=mitigation,
+        file_path=vulnerability.get('source_comp_id'),
+        severity_justification="CVSS v3 base score: {}\nCVSS v2 base score: {}".format(cvss_v3, cvss_v2),
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
        severity_justification=f"CVSS v3 base score: {cvss_v3}\nCVSS v2 base score: {cvss_v2}",
```

Less verbose format string; [see docs here](https://docs.python.org/3/reference/lexical_analysis.html#f-strings). New in Python 3.6.

File: dojo/tools/jfrogxray/parser.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        title = str(cve) + " - " + vulnerability['component']
+
+    # create the finding object
+    finding = Finding(
+        title=title,
+        cve=cve,
+        cwe=cwe,
+        test=test,
+        severity=severity,
+        description=vulnerability['summary'] + "\n\n" + extra_desc,
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
        description=(vulnerability['summary'] + "\n\n" + extra_desc).strip(),
```

File: dojo/tools/jfrogxray/parser.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        mitigation=mitigation,
+        file_path=vulnerability.get('source_comp_id'),
+        severity_justification="CVSS v3 base score: {}\nCVSS v2 base score: {}".format(cvss_v3, cvss_v2),
+        static_finding=True,
+        dynamic_finding=False,
+        references=vulnerability.get('component_versions').get('more_details').get('provider'),
+        impact=severity)
+
+    finding.description = finding.description.strip()
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
```

Combined with the move suggestion.

File: dojo/tools/jfrogxray/parser.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        items = {}
+        if 'data' in tree:
+            vulnerabilityTree = tree['data']
+
+            for node in vulnerabilityTree:
+
+                item = get_item(node, test)
+
+                title_cve = "No CVE"
+                if 'cves' in node.get('component_versions').get('more_details'):
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
                more_details = node.get('component_versions').get('more_details')
                if 'cves' in more_details:
```

File: dojo/tools/jfrogxray/parser.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                    if 'cve' in node.get('component_versions').get('more_details').get('cves')[0]:
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
                    cve = more_details.get('cves')[0]
                    if 'cve' in cve:
```

File: dojo/tools/jfrogxray/parser.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                        title_cve = node.get('component_versions').get('more_details').get('cves')[0].get('cve')
```
Comment (jvz) [inline dojo/tools/jfrogxray/parser.py]:
```suggestion
                        title_cve = cve.get('cve')
```

File: dojo/tools/jfrogxray/parser.py (hunk 1 above)
Comment (madchap) [reply dojo/tools/jfrogxray/parser.py]: My bad, it should not even be here. This is when I was quickly debugging. I will remove it.

File: dojo/tools/jfrogxray/parser.py (hunk 6 above)
Comment (madchap) [reply dojo/tools/jfrogxray/parser.py]: Thanks, will do that.

**Comment by madchap** [comment]
Travis build issue for Docker, with mysql-client package not being found. cc @Maffooch @dr3dd589

**Comment by Maffooch** [comment]
The docker breakage from this weekend has been fixed. Please pull the fix and push again for travis to run. Could you also submit sample scans to the sample-scans repo as well please? I will approve after the test passes. Good work on implementing the parser.

**Comment by madchap** [comment]
Hi @Maffooch 

Unfortunately, I can't submit sample scans. I do not run the tool myself, got some json reports that contain information about proprietary product. I will do whenever I get a chance to submit something more "open".

Thanks,

PS: Looks like Travis did not even attempt to run.

**Comment by Maffooch** [comment]
@madchap I was referring to the sample scans used to test the parser. The ones with names like one_vuln, many_vuln, etc. 

I believe travis didn't bother to run because of the conflicts. A lot of stuff has been merged today.

**Comment by madchap** [comment]
> @madchap I was referring to the sample scans used to test the parser. The ones with names like one_vuln, many_vuln, etc.

They are in the first commit (one_vuln and many_vulns).

**Comment by Maffooch** [comment]
@madchap Right. Could you submit them [here](https://github.com/DefectDojo/sample-scan-files) as well please?

**Comment by Maffooch** [comment]
Hey @madchap could you fix merge conflicts so that this could be approved? Great work on the parser.

**Pull request merged**